infra/relayer/src/validator.js

```diff
@@ -0,0 +1,240 @@
+    if (method == undefined) {
+      logger.info(
+        `Validation failed. Method ${signature} not found on ${this.name} ${this.address}`
+      )
+      return false
+    }
+    logger.info(`  Transaction calls ${method.name} on ${this.name}`)
+    const paramsHex = txdata.substr(8)
+    let params = {}
+    try {
+      params = web3.eth.abi.decodeParameters(method.inputs, paramsHex)
+    } catch (err) {
+      // Do nothing, validateMethod can params data that can't be decoded.
+    }
+    return this.validateMethod(method, txdata, params)
+  }
+  /**
+   * validateMethod is intended to be overwritten for invdividual contract types.
+   * @param {object} method
+   * @param {string} txdata
+   * @param {object} params
+   */
+
+  // eslint-disable-next-line no-unused-vars
+  validateMethod(method, txdata, params) {
+    // Empty list of new validations to be added
+    // means that no further validation is required for this method
+    return []
+  }
+
+  isValidTokenAddress(address) {
+    const validTokens = [this.addresses.OGN, this.addresses.DAI]
+    for (const token of validTokens) {
+      if (
+        address.toLowerCase().replace(/^0x/, '') ===
+        token.toLowerCase().replace(/^0x/, '')
+      ) {
+        return true
+      }
+    }
+    return false
+  }
+
+  _methodsBySignature(jsonInterface) {
+    const methods = {}
+    jsonInterface
+      .filter(i => i.type === 'function' && !i.constant)
+      .forEach(o => (methods[o.signature.replace(/^0x/, '')] = o))
+    return methods
+  }
+}
+
+class ProxyValidator extends ContractCallVailidator {
+  // eslint-disable-next-line no-unused-vars
+  validateMethod(method, txdata, params) {
+    const validationsNeeded = []
+    const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
+
+    if (
+      method.name == 'marketplaceExecute' ||
+      method.name == 'transferTokenMarketplaceExecute'
+    ) {
+      if (!this.isValidTokenAddress(params._token)) {
+        logger.info(
+          `Validation failed. ${params._token} is not an allowed token address`
+        )
+        return false
+      }
+      validationsNeeded.push([params._marketplace, params._offer])
+    } else if (method.name == 'marketplaceFinalizeAndPay') {
+      const isEth = params._currency === ZERO_ADDRESS
+      if (!(isEth || this.isValidTokenAddress(params._currency))) {
+        logger.info(
+          `Validation failed. ${params._currency} is not an allowed currency address`
+        )
+        return false
+      }
+      validationsNeeded.push([params._marketplace, params._finalize])
+    } else if (method.name == 'swapAndMakeOffer') {
+      if (!this.isValidTokenAddress(params._token)) {
+        logger.info(
+          `Validation failed. ${params._token} is not an allowed ERC20 address`
+        )
+        return false
+      }
+      validationsNeeded.push([params._marketplace, params._offer])
+      validationsNeeded.push([params._exchange, params._swap])
+    } else {
```

Comment:
Do we want to limit this to only these 4 methods?  Just thinking that we were talking about adding `transferToOwner()` for sweeps yesterday.

Reply:
I've added `transferToOwner` to the list. If there's anything else we need later, we can add it then.